### PR TITLE
Fix use-before-def errors for ESNext property declarations

### LINKED ISSUE #36441: Regression: property used before initialization inside function
**TypeScript Version:**  3.8.0-dev.20200125

**Search Terms:**

**Code**

```ts
// @target: esnext
// @useDefineForClassFields. true
class C {
    bar = () => this.foo1 + this.foo2;
    foo1 = '';
    constructor(public foo2: string) {}
}
```

**Expected behavior:**

Accessing `foo1` and `foo2` should not be an error as the function is not immediately invoked and therefore the properties are initialized when accessed. This worked before #36425.

**Actual behavior:**

used before initialization error on `this.foo1` and `this.foo2`

**Playground Link:** https://www.typescriptlang.org/play/?useDefineForClassFields=true&target=99&ts=3.8.0-dev.20200125&ssl=9&ssc=2&pln=5&pc=2#code/MYGwhgzhAEDC0G8BQ1XQEZgE7QLzQAoBKPAPmgBcALASwgDoAzAe2YEZoBqS2hl5gEwBuFGn4d8Ackki00YMwB2EClgCuwCsywEADmvQgawaPwEAuaCqw1FAcxIIAvkidA

**Related Issues:** #36425 @sandersn

### LINKED ISSUE #36442: ESNext+[[Define]]: false positive error accessing parameter property of base class in initializer of derived class
**TypeScript Version:**  3.8.0-dev.20200125

**Search Terms:**

**Code**

```ts
// @target: esnext
// @useDefineForClassFields. true
class C {
    constructor(public foo: string) {}
}

class D extends C {
    bar = this.foo;
}
```

**Expected behavior:**

No error. The constructor of `C` (and therefore the initialization of `foo`) is already executed when `bar` is initialized.

**Actual behavior:**

used before initialization error on `this.foo`

**Playground Link:** https://www.typescriptlang.org/play/?useDefineForClassFields=true&target=99&ts=3.8.0-dev.20200125&ssl=7&ssc=2&pln=1&pc=1#code/MYGwhgzhAEDC0G8BQ1XWAewHYQC4CcBXYXDfACgAdCAjEAS2GgDMMMAuaPferAcwCUiAL5JRSUJBgARaAFMAHrjlYAJjHjI00GmHzQAvNFwALehAB0rDAG4xQA

**Related Issues:** #36425 @sandersn

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #36441
Fixes #36442

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1326,12 +1326,14 @@ namespace ts {
                 }
                 else if (isPropertyDeclaration(declaration)) {
                     // still might be illegal if a self-referencing property initializer (eg private x = this.x)
-                    return !isPropertyImmediatelyReferencedWithinDeclaration(declaration, usage);
+                    return !isPropertyImmediatelyReferencedWithinDeclaration(declaration, usage, /*stopAtAllPropertyDeclarations*/ false);
                 }
                 else if (isParameterPropertyDeclaration(declaration, declaration.parent)) {
+                    const container = getEnclosingBlockScopeContainer(declaration.parent);
                     // foo = this.bar is illegal in esnext+useDefineForClassFields when bar is a parameter property
                     return !(compilerOptions.target === ScriptTarget.ESNext && !!compilerOptions.useDefineForClassFields
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: For some reason my brain wants this expression to be De Morgan’d, or maybe for whole function itself to return use-before-declared, not declared-before-used. Oh well.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1357,10 +1359,19 @@ namespace ts {
             }
 
             const container = getEnclosingBlockScopeContainer(declaration);
-            return !!(usage.flags & NodeFlags.JSDoc)
-                || isInTypeQuery(usage)
-                || isUsedInFunctionOrInstanceProperty(usage, declaration, container)
-                   && !(compilerOptions.target === ScriptTarget.ESNext && !!compilerOptions.useDefineForClassFields);
+            if (!!(usage.flags & NodeFlags.JSDoc) || isInTypeQuery(usage)) {
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: I bet this condition actually could be the same as `isValidTypeOnlyAliasUseSite`. But probably not worth changing.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1427,8 +1438,9 @@ namespace ts {
 
                     switch (node.kind) {
                         case SyntaxKind.ArrowFunction:
-                        case SyntaxKind.PropertyDeclaration:
                             return true;
+                        case SyntaxKind.PropertyDeclaration:
+                            return stopAtAllPropertyDeclarations ? "quit": true;
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: I can’t quite figure out what this means without stepping through one case that needs `stopAtAllPropertyDeclarations` as `true` and one that needs `false`, but are there any tests with a class expression as the initializer of a property declaration?

```ts
class Foo {
  Bar = class extends Foo {
    p2 = this.p1
  }
  p1 = 0
}
```

(I don’t know, maybe there’s a more creative combo of these horrors)

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts]: the WHOLE time I was working on this, except every change made things harder to read.

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [reply src/compiler/checker.ts]: There is one: 

```ts
class B2<V> {
    anon = class extends A<V> { }
}
```

Your example works on 3.7, so needs to keep working on 3.8:

https://www.typescriptlang.org/play/index.html#code/MYGwhgzhAECC0G8BQ1XQELQLzVJGApgB4AuBAdgCYzzJr3QAOATNtCQBYCWEAdIwEYUaAL7DUgtgAYkYgG5gATtDBtyBAO5wAFAEokwAPbkIhkAV4hDAc21h+A-QuVgARms0re6PQeOnzSxs7Vwd9IA

Back to the drawing board I guess. What a test case!

Comment (andrewbranch) [reply src/compiler/checker.ts]: Commit to fix this looks good 👍

**Pull request merged**